lms/djangoapps/instructor_task/tests/test_tasks_helper.py

```diff
@@ -1801,7 +1816,8 @@ def test_grade_report_with_overrides(self):
                         'Homework 1: Subsection': '0.5',
                         'Homework 2: Unattempted': '1.0',
                         'Homework 3: Empty': 'Not Attempted',
-                        'Homework (Avg)': str(3.0 / 6.0),
+                        'Homework 4: Unreleased': 'Not Attempted',
+                        'Homework (Avg)': str(1.5 / 4),
```

Comment:
Is this a user-visible change? i.e. will users see different grades or items listed in the grade report now, if their course has graded questions in an upcoming unreleased section? I hope the answer is no, but just want to confirm.

Reply:
Great question!

That is a potential issue because we're changing the transformers used to produce the user specific structure;
in the case of the grade report instructor task (and some other places I've looked) it won't be an issue because the list of headers is derived from the collected structure, i.e., it's already assuming it needs to look at everything.
However, **if there is some code that generates the headers from the user-specific structure**, that code will see unreleased sections that it did not see before.

An alternative fix here would be to change `CourseGrade.subsection_grade` to try using `CourseData.structure` first (because it might have user overrides) and falling back to `CourseData.collected_structure`.
I think this is what `CourseData.effective_structure` is attempting but failing to do and would delegate the access control decision to calling code, which is more appropriate IMO.

Reply:
Sorry for the slow reply. I feel a little out of my depth here as I'm not too familiar with this part of the code. 

> An alternative fix here would be to change CourseGrade.subsection_grade to try using CourseData.structure first (because it might have user overrides) and falling back to CourseData.collected_structure.

I'm hesitant to change the behavior of the whole transformer cache that produces the user-specific structure. The alternative fix you're suggesting makes more sense to me, and seems to align better with [the docstring of `subsection_grade`](https://github.com/edx/edx-platform/blob/28e5812637c9d9995c1c34c3ccca3c3e5760bf09/lms/djangoapps/grades/course_grade.py#L58-L61) anyway.

It's unfortunate that none of these `CourseData` methods have any docstrings, because it's hard to even tell what they're supposed to be doing. As you point out, the behavior of `CourseData.effective_structure` changes depending on whether or not `.structure` has been accessed, and at first I didn't know if that was a feature, a confusing design, or a bug. Without a clear explanation of what it's supposed to be doing, it was hard to say.

That said, I did notice that [this is the only use of `effective_structure` outside of the `CourseData` class](https://github.com/edx/edx-platform/search?q=effective_structure), and now that I've looked more closely at the code, I _think_ I know the intention of `effective_structure`. I would suggest updating your PR to add this docstring to `effective_structure`:

```python
    @property
    def effective_structure(self):
        """
        Get whichever course block structure is already loaded, if any.
        
        This may give either the user-specific course structure or the generic
        structure, depending on which is cached at the moment. Because of that,
        this should only be used for queries related to the root block of the
        course, which will always exist in either structure.

        For anything else, such as queries involving course sections or blocks,
        use either .structure or .collected_structure to explicitly state
        whether you want the user-specific version of the course or not.
        """
        return self._structure or self._collected_block_structure
```

.. and then implementing the alternative fix you've suggested.

Reply:
@bradenmacdonald Done!